src/main/java/org/kiwiproject/dropwizard/util/startup/StartupLocker.java

```diff
@@ -0,0 +1,136 @@
+package org.kiwiproject.dropwizard.util.startup;
+
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+import static org.kiwiproject.base.KiwiPreconditions.requireNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.dropwizard.setup.Environment;
+import io.dropwizard.util.Duration;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.locks.InterProcessLock;
+import org.apache.curator.framework.recipes.locks.InterProcessMutex;
+import org.kiwiproject.curator.CuratorFrameworkHelper;
+import org.kiwiproject.curator.CuratorLockHelper;
+import org.kiwiproject.curator.config.CuratorConfig;
+import org.kiwiproject.curator.exception.LockAcquisitionException;
+import org.kiwiproject.curator.zookeeper.ZooKeeperAvailabilityChecker;
+import org.kiwiproject.dropwizard.util.startup.PortAssigner.PortAssignment;
+import org.kiwiproject.dropwizard.util.startup.listener.StartupJettyLifeCycleListener;
+import org.kiwiproject.dropwizard.util.startup.listener.StartupWithLockJettyLifeCycleListener;
+
+import java.util.Optional;
+
+/**
+ * Utilities to acquire and release a lock from ZooKeeper during startup of the service
+ */
+@Slf4j
+public class StartupLocker {
+
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private final ZooKeeperAvailabilityChecker zkAvailabilityChecker;
+
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private final CuratorFrameworkHelper curatorFrameworkHelper;
+
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private final CuratorLockHelper curatorLockHelper;
+
+    private final SystemExecutioner executioner;
+
+    @Builder
+    private StartupLocker(SystemExecutioner executioner,
+                          ZooKeeperAvailabilityChecker zkAvailabilityChecker,
+                          CuratorFrameworkHelper curatorFrameworkHelper,
+                          CuratorLockHelper curatorLockHelper) {
+        this.executioner = requireNotNull(executioner, "SystemExecutioner is required");
+        this.zkAvailabilityChecker = Optional.ofNullable(zkAvailabilityChecker).orElse(new ZooKeeperAvailabilityChecker());
+        this.curatorFrameworkHelper = Optional.ofNullable(curatorFrameworkHelper).orElse(new CuratorFrameworkHelper());
+        this.curatorLockHelper = Optional.ofNullable(curatorLockHelper).orElse(new CuratorLockHelper());
+    }
+
+    @AllArgsConstructor
+    public static class StartupLockInfo {
```

Comment:
I can't remember, but if this is intended for use outside this library, then it should be documented, at least at the top level.

Reply:
Our base application class uses this object, so I will document it.